src/netif.rs

```diff
@@ -952,6 +949,154 @@ mod driver {
+        // Shall we remove SLIP support?
+        #[cfg(esp_idf_lwip_slip_support)]
+        pub fn new_nonstatic_slip<'d, F>(netif: T, tx: F) -> Result<Self, EspError>
+        where
+            F: FnMut(&[u8]) -> Result<(), EspError> + Send + 'd,
+        {
+            Self::new_nonstatic(netif, None, None, |_| Ok(()), tx)
+        }
+
+        /// Create a new netif driver around the provided `EspNetif` instance
+        /// and the TX callback that the driver would call when it wants to ingest
+        /// a packet into the underlying transport:
+        /// * The transport can be anything, but with - say - PPP netif - it would
+        ///   typically be UART, and the callback implementation is simply expected
+        ///   to write the PPP packet into UART
+        /// * `got_ip_event_id` and `lost_ip_event_id` are the event IDs that the driver
+        ///   will listen to so that it can connect/disconnect the netif upon receival
+        ///   / loss of IP
+        /// * `post_attach_cfg` is a netif-specific configuration that will be executed
+        ///   after the netif is attached. See `ppp_config` below for a PPP-specific driver
+        ///   post-attach configuration
+        ///
+        /// # Safety
+        ///
+        /// This method - in contrast to method `new` - allows the user to pass
+        /// non-static callbacks/closures. This enables users to borrow
+        /// - in the closure - variables that live on the stack - or more generally - in the same
+        ///   scope where the service is created.
+        ///
+        /// HOWEVER: care should be taken NOT to call `core::mem::forget()` on the service,
+        /// as that would immediately lead to an UB (crash).
+        /// Also note that forgetting the service might happen with `Rc` and `Arc`
+        /// when circular references are introduced: https://github.com/rust-lang/rust/issues/24456
+        ///
+        /// The reason is that the closure is actually sent to a hidden ESP IDF thread.
+        /// This means that if the service is forgotten, Rust is free to e.g. unwind the stack
+        /// and the closure now owned by this other thread will end up with references to variables that no longer exist.
+        ///
+        /// The destructor of the service takes care - prior to the service being dropped and e.g.
+        /// the stack being unwind - to remove the closure from the hidden thread and destroy it.
+        /// Unfortunately, when the service is forgotten, the un-subscription does not happen
+        /// and invalid references are left dangling.
+        ///
+        /// This "local borrowing" will only be possible to express in a safe way once/if `!Leak` types
+        /// are introduced to Rust (i.e. the impossibility to "forget" a type and thus not call its destructor).
+        pub fn new_nonstatic<F, P>(
+            netif: T,
+            got_ip_event_id: Option<i32>,
```

Comment:
if you are talking about ip_event_id's you are talking about [this](https://github.com/espressif/esp-idf/blob/master/components/esp_netif/include/esp_netif_types.h#L98) right? Why do we accept a i32 here and not one of this concrete variants?

Reply:
because `ip_event_t` is anyway defined as `pub type ip_event_t = ::core::ffi::c_uint (= u32);` by `bindgen`. A C-style enum is not converted to a Rust enum with our current `bindgen` settings, and long ago we decided not to touch this (it is like that by default in `bindgen`).

Reply:
To be clear, i mainly talking about the ergonomics of the API. The user see's that this api wants a u32 or something, but now needs to find something in the bindgen that is an ip_event_t_* to get the actual thing. So there is a slightly disconnect here.

Reply:
What would you suggest?
`NonZeroIp_event_t` does not exist and if I replace `Option<NonZeroU32>` with `Option<ip_event_t>` it is not so perfect either, as user then might pass Some(0), which does not make sense, as it is the same as `None`.

It should be either just `ip_event_t` or `Option<NonZeroU32>` and each has pros and cons.

Reply:
Ah ok the code example made it more clear. So we essentially use just some of the NetifConfiguration parameters here so the user dont need to provide the NonZerou32 from thin air.

Wouldn't it in that case be a bit more nicer to use for somebody who just look at this, if we accept for example an &NetifConfiguration and internally extract the id? 
 
if we in the example end up promoting to passing it from a netifconfiguration, why not clearly communicate it in the methods argument type?

Reply:
Even passing `NetifConfiguration` is dirty, because you might pass one netif configuration to the `EspNetif` you are wrapping, while another - to the `EspNetifDriver` that wraps the netif.

Look now. The got_ip/lost_ip are extracted **from** the `EspNetif` now. I.e. they are the same that are used for configuring the netif, as it also needs to know about those :)
`EspNetif` is no longer just a handle, but how many netifs is one going to have? Definitely not a thousand for this to be a concern.